common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java

```diff
@@ -225,76 +246,137 @@ public String encrypt(final String clearText)
 
         final String encryptedText = new String(Base64.encode(blobVerAndEncryptedDataAndIVAndMacDigest,
                 Base64.NO_WRAP), AuthenticationConstants.ENCODING_UTF8);
-        Log.v(TAG, "Finished encryption");
+        Logger.verbose(TAG + methodName, "Finished encryption");
 
         return getEncodeVersionLengthPrefix() + ENCODE_VERSION + encryptedText;
     }
 
     @Override
     public String decrypt(final String encryptedBlob) throws GeneralSecurityException, IOException {
-        final String methodName = "decrypt";
-        Log.v(TAG, "Starting decryption");
+        final String methodName = ":decrypt";
+        Logger.verbose(TAG + methodName, "Starting decryption");
 
         if (StringExtensions.isNullOrBlank(encryptedBlob)) {
             throw new IllegalArgumentException("Input is empty or null");
         }
 
-        int encodeVersionLength = encryptedBlob.charAt(0) - 'a';
-        validateEncodeVersion(encryptedBlob, encodeVersionLength);
+        if (getEncryptionType(encryptedBlob) == EncryptionType.UNENCRYPTED) {
+            Logger.warn(TAG + methodName, "This string is not encrypted. Finished decryption.");
+            return encryptedBlob;
+        }
 
-        final byte[] bytes = Base64.decode(
-                encryptedBlob.substring(1 + encodeVersionLength),
-                Base64.DEFAULT
-        );
+        final String packageName = getPackageName();
+        final List<KeyType> keysForDecryptionType = getKeysForDecryptionType(encryptedBlob, packageName);
+
+        final byte[] bytes = getByteArrayFromEncryptedBlob(encryptedBlob);
+        for (final KeyType keyType : keysForDecryptionType) {
+            try {
+                final SecretKey secretKey = loadSecretKey(keyType);
+                if (secretKey == null) {
+                    continue;
+                }
+
+                String result = decryptWithSecretKey(bytes, secretKey);
+                Logger.verbose(TAG + methodName, "Finished decryption.");
+                return result;
+            } catch (GeneralSecurityException | IOException e) {
+                Logger.error(
+                        TAG + methodName,
+                        "Failed to decrypt with KeyType: "
+                                + keyType.toString(),
+                        e
+                );
+            }
+        }
+
+        Logger.verbose(
+                TAG + methodName,
+                "Tried all decryption keys and decryption still fails. Throw an exception.");
+
+        throw new GeneralSecurityException(ErrorStrings.DECRYPTION_FAILED);
```

Comment:
>DECRYPTION_FAILED [](start = 56, length = 17)

I know this is a late ask... I think it is worthwhile to have telemetry on all failed decryption attempts - these instances should go down and if they are not going down, then we know we have a bug.

Reply:
already bubbling this up in part 2.